Correct ElectronDiffraction2D.set_diffraction_calibration()

emichr opened the pull request:
---
name: Correct ElectronDiffraction2D.set_diffraction_calibration()
about: Correcting code in set_electron_diffraction_calibration() to match its docstring.

---

**Checklist**
- [ ] Updated CHANGELOG.md
- [ ] Marked as finished

**What does this PR do? Please describe and/or link to an open issue.**
This code fixes a bug in the ElectronDiffraction2D.set_diffraction_calibration() function. When the `center` argument was given, this function used to set the `offset` attribute of the diffraction axes in pixel units and not in the new calibrated units. With the new changes, the code matches what is described in the docstring and its intended use.

CSSFrancis commented:
@emichr can you please rebase this PR on the `main` branch just to make sure all of the tests are passing?

emichr commented:
Hm, it still looks like the tests are failing @CSSFrancis. It seems the tests assumes the center passed to the function is in calibrated units, not in pixels. Perhaps the best is then to change the docstring, as the suggested bugfix may break other functionality?

CSSFrancis commented:
> Hm, it still looks like the tests are failing @CSSFrancis. It seems the tests assumes the center passed to the function is in calibrated units, not in pixels. Perhaps the best is then to change the docstring, as the suggested bugfix may break other functionality?

I would change the doc strings then!  If you want to use pixel units then you can always add the argument `pixel_units` and then add an if statement.  That might be a nice feature to add.

emichr commented:
> > Hm, it still looks like the tests are failing @CSSFrancis. It seems the tests assumes the center passed to the function is in calibrated units, not in pixels. Perhaps the best is then to change the docstring, as the suggested bugfix may break other functionality?
> 
> I would change the doc strings then! If you want to use pixel units then you can always add the argument `pixel_units` and then add an if statement. That might be a nice feature to add.

I added the `center_pixels` keyword instead as I thought that would be clearer - please let me know if you disagree. I modified the relevant tests as well to include the new keyword. When looking at the tests, I also noticed that the unit test didn't actually check what was being done when no center was given as an argument, so I threw that change in there as well. Please let me know if there's something off!

emichr commented:
Nevermind @CSSFrancis, I've overlooked some stuff - I'll look into it and fix it as soon as I have the time and notify you when I've sorted it. Sorry for the back-and-forth on this simple issue.

emichr commented:
I got the time sooner than I thought, so I think it should be alright now. Please let me know if there's anything missing @CSSFrancis (or @magnunor if you have the time).

CSSFrancis commented:
@emichr any chance you can rebase this again and then we can get this merged assuming that the CI passes?

emichr commented:
@CSSFrancis I squashed the commits and rebased the branch - hope everything works out and that all looks good! Let me know if there are anything that needs fixing.

CSSFrancis commented on pyxem/signals/electron_diffraction2d.py:
```diff
@@ -175,7 +175,7 @@ def set_experimental_parameters(
             )
 
-    def set_diffraction_calibration(self, calibration, center=None):
+    def set_diffraction_calibration(self, calibration, center=None, center_pixels=None):
```
```suggestion
    def set_diffraction_calibration(self, calibration, center=None, pixel=False):
```

emichr replied:
Using a `bool` is a good suggestion, I looked into it earlier when you suggested it. I tried the other solution as I found that to be a little bit clearer in terms of argument names and implementation. If we want to go with a boolean argument, I think users might find `pixels` a bit counterintuitive (it can be interpreted in many ways). Of course the docstring explains the usage, but perhaps we can come up with a better argument name? What do you think? 

An alternative might be a type check as is used in Hyperspy? If two `float`s are passed as the center argument it is interpreted as given in calibrated units, while a pair of `int`s are interpreted as pixel units?

emichr replied:
> An alternative might be a type check as is used in Hyperspy?

With this I think about `signal.inav[...]`, etc, usage by the way.

emichr replied:
I agree, doing this is not completely straight forward. I like the idea of splitting it into two separate methods. The old `set_diffraction_calibration` could in principle be continued where it applies these two methods and the docstring could easily describe the order of operations etc. If you like, I could try and see if I can come up with a suggestion for it? However, I am a bit short on time and I will be away on paternity leave for a few months soon so I might go AWOL without much notice.

ericpre replied:
> I was kind of hoping [hyperspy/hyperspy#2756](https://github.com/hyperspy/hyperspy/pull/2756) would be merged into hyperspy and that might fix this problem too :)

@CSSFrancis, yes, we should get this PR sorted along with https://github.com/hyperspy/hyperspy/pull/3385.
xref https://github.com/hyperspy/hyperspy/pull/2756#issuecomment-2169287069

CSSFrancis commented on pyxem/signals/electron_diffraction2d.py:
```diff
@@ -184,11 +184,19 @@ def set_diffraction_calibration(self, calibration, center=None):
-            the data array is assumed to be the center of the pattern.
+            Position of the direct beam center, in calibrated units. If None the center of
+            the data array is assumed to be the center of the pattern unless `center_pixels` is given.
+        center_pixels : tuple
+            Position of the direct beam center, in pixels. Will only be used if `center` is not given.
```
```suggestion
        pixels : bool
            If True, the provided center is in pixel coordinates. Default is False which means
            the provided center is in calibrated coordinates.
```

emichr replied:
Sorry @CSSFrancis, I suddenly got tied up with parental leave - but I'll take a look and commit your suggestion and the related change in the unit tests this week!

CSSFrancis commented on pyxem/signals/electron_diffraction2d.py:
```diff
@@ -184,11 +184,19 @@ def set_diffraction_calibration(self, calibration, center=None):
+            Position of the direct beam center, in pixels. Will only be used if `center` is not given.
         """
-        if center is None:
+        if center is None and center_pixels is None:  # Use center of the pattern array
             center = np.array(self.axes_manager.signal_shape) / 2 * calibration
+        elif center is None and center_pixels is not None:
+            center = (
+                np.array(center_pixels) * calibration
+            )  # Use provided center in pixel coordinates
+        else:
+            pass  # Ignore `center_pixels` if `center` is provided.
```
```suggestion
        if center is None:  # Use center of the pattern array
            center = np.array(self.axes_manager.signal_shape) / 2 * calibration
        elif center is not None and pixels:  # Use provided center in pixel coordinates
            center = (
                np.array(center) * calibration
            )  # Use provided center in pixel coordinates
        else:
            center = np.array(center)  # Use provided center in calibrated coordinates
```

CSSFrancis reviewed (changes requested):
@emichr sorry I've been pretty bad at reviewing this. :)

Just a couple of suggestions.  I think it makes more sense to have a bool argument rather than allow people to pass a `center` and a `center_pixels` argument and have it be not very clear about which is being used.

CSSFrancis commented on pyxem/tests/signals/test_electron_diffraction2d.py:
```diff
@@ -134,27 +134,23 @@ def test_set_scan_calibration(self, diffraction_pattern):
         assert isinstance(diffraction_pattern, ElectronDiffraction2D)
 
     @pytest.mark.parametrize(
```
This could be simplified a little bit with the change.

Something like:
```python
    @pytest.mark.parametrize(
        "calibration, center, pixels",
        [
            (0.5, (4, 4), True),
            (0.5, (2, 2), False),
            (0.5, None, False),
        ],
    )
    def test_set_diffraction_calibration(
        self, diffraction_pattern, calibration, center, pixels
    ):
        diffraction_pattern.set_diffraction_calibration(
            calibration, center=center, pixels=pixels
        )
        dx, dy = diffraction_pattern.axes_manager.signal_axes
        assert dx.scale == calibration and dy.scale == calibration
        assert dx.offset == -2 and dy.offset == -2
```

emichr replied:
I think the test will need to be a little more complex than this. For instance, `dx.offset` will depend on the `center` and `pixels` values. In addition, I think it is a good idea to also check on the data values and make sure that the specified center matches what is retrieved on the data side when accessed through `dp.isig[...]`

CSSFrancis commented:
@emichr hopefully this helps and makes sense!  

@hakonanes any chance you can add me as a maintainer to pyxem. It would be nice to be able to direct edit pull requests. :)

ericpre commented:
> Edit: That might actually be because of a weird GitHub authentication error

Yes, I think that I had this issue before but I don't know more than that...
